File: interpretation-ui/src/main/java/com/gcgenome/lims/dto/InterpretationPanel.java
package com.gcgenome.lims.dto;

import jsinterop.annotations.JsPackage;
import jsinterop.annotations.JsProperty;
import jsinterop.annotations.JsType;

import java.util.List;

@JsType(isNative = true, namespace= JsPackage.GLOBAL, name="Object")
public class InterpretationPanel {
    public String result;
    @JsProperty(name="result_text")
    public String resultText;
    @JsProperty(name="reason_for_referral")
    public String reasonForReferral;
    @JsProperty(name="clinical_information")
    public String clinicalInformation;
    @JsProperty(name="abbreviation_reference")
    public String abbreviationReference;
    @JsProperty(name="abbreviation_disease")
    public String abbreviationDisease;
    public String abbreviation;
    public String interpretation;
    @JsProperty(name="mean_depth")
    public String meanDepth;
    public String coverage;
    public Variant[] variants;
    public String recommendation;
    public InterpretationPanel addendum;
    @JsProperty(name="incidental_findings")
    public InterpretationPanel incidentalFindings;
    public List<Author> authors;
    public Boolean revision;
    @JsProperty(name="mlpa_result")
    public Mlpa mlpaResult;
    @JsType(isNative = true, namespace= JsPackage.GLOBAL, name="Object")
    public static class Variant {
        public String snv;
        public String analysis;
        public String gene;
        public String hgvsc;
        public String hgvsp;
        @JsProperty(name="origin_hgvsc")
        public String originHgvsc;
        @JsProperty(name="origin_hgvsp")
        public String originHgvsp;
        public String zygosity;
        public String disease;
        @JsProperty(name="disease_full_name")
        public String diseaseFullName;
        public String inheritance;
        @JsProperty(name="class")
        public String clazz;
        public String interpretation;
        @JsProperty(name="gen_phen_db")
        public String genPhenDb;

        // for client only
        public String mim;
        public List<VariantRaw.ReportInfo> reported;
    }
    @JsType(isNative = true, namespace= JsPackage.GLOBAL, name="Object")
    public static class Author {
        public String name;
        public String comment;
    }
    @JsType(isNative = true, namespace= JsPackage.GLOBAL, name="Object")
    public static class Mlpa {
        public String gene;
        public String result;
        public String exons;
        public String zygosity;
        @JsProperty(name="del_dup")
        public String delDup;
    }
}

File: interpretation/src/main/kotlin/com/gcgenome/lims/dto/InterpretationPanel.kt
package com.gcgenome.lims.dto

import com.fasterxml.jackson.annotation.JsonProperty
import com.gcgenome.lims.constants.Clazz
import java.util.*
import java.util.stream.Collectors

data class InterpretationPanel (
    val result: String?,
    val resultText: String?,
    val reasonForReferral: String?,
    val clinicalInformation: String?,
    val abbreviationReference: String?,
    val abbreviationDisease: String?,
    val abbreviation: String?,
    val interpretation: String?,
    val meanDepth: String?,
    val coverage: String?,
    val variants: List<Variant>,
    val recommendation: String?,
    val addendum: InterpretationPanel?,
    val incidentalFindings: InterpretationPanel?,
    val authors: List<Author> = emptyList(),
    val revision: Boolean = false
) {
    open class InterpretationPanelBuilder {
        var result: String? = null
        var resultText: String? = null
        var reasonForReferral: String? = null
        var clinicalInformation: String? = null
        private val abbreviationReference = Collections.synchronizedList(ArrayList<String>())
        private val abbreviationDisease = Collections.synchronizedList(ArrayList<String>())
        private val abbreviationInheritance = Collections.synchronizedList(ArrayList<String>())
        private val abbreviationZygosity = Collections.synchronizedList(ArrayList<String>())
        private val abbreviationClass = Collections.synchronizedList(ArrayList<String>())
        var interpretationPrefix: String? = null
        private val interpretation = Collections.synchronizedList(ArrayList<String>())
        var meanDepth: String? = null
        var coverage: String? = null
        val variants = Collections.synchronizedList(ArrayList<Variant>())
        var recommendation: String? = null
        var addendum: InterpretationPanel? = null
        var incidentalFindings: InterpretationPanel? = null
        var authors: List<Author> = emptyList()
        var revision: Boolean = false

        fun appendReference(abbreviationReference: String?)             = this.also { if (abbreviationReference != null) this.abbreviationReference.add(abbreviationReference) }
        fun appendDisease(abbreviationDiseases: List<String>)           = this.also { if (abbreviationDiseases != null) this.abbreviationDisease.addAll(abbreviationDiseases) }
        fun appendInheritance(abbreviationInheritances: List<String>)   = this.also { if (abbreviationInheritances != null) this.abbreviationInheritance.addAll(abbreviationInheritances) }
        fun appendZygosity(abbreviationZygosity: String?)               = this.also { if (abbreviationZygosity != null) this.abbreviationZygosity.add(abbreviationZygosity) }
        fun appendClass(abbreviationClass: String?)                     = this.also { if (abbreviationClass != null) this.abbreviationClass.add(abbreviationClass) }
        fun appendInterpretation(interpretation: String?)               = this.also { if (interpretation != null) this.interpretation.add(interpretation) }
        fun appendVariant(variant: Variant?)                            = this.also { if (variant != null) variants.add(variant) }
        fun build(): InterpretationPanel {
            val abbreviationReference = abbreviationReference.stream().distinct().filter(String::isNotBlank).distinct().collect(Collectors.joining("; "))
            val abbreviationDisease = abbreviationDisease.stream().distinct().filter(String::isNotBlank).distinct().collect(Collectors.joining("; "))
            val abbreviation = (abbreviationZygosity + abbreviationInheritance + abbreviationClass).stream().filter(String::isNotBlank).distinct().collect(Collectors.joining("; "))
            val interpretation = (if(interpretationPrefix!=null) "$interpretationPrefix\n\n" else "") + interpretation.stream().distinct().filter(String::isNotBlank).collect(Collectors.joining("\n\n"))
            return InterpretationPanel(
                result=result, resultText=resultText, reasonForReferral=reasonForReferral, clinicalInformation=clinicalInformation,
                abbreviationReference=abbreviationReference, abbreviationDisease=abbreviationDisease, abbreviation=abbreviation,
                interpretation=interpretation,
                meanDepth=meanDepth, coverage=coverage,
                variants=variants,
                recommendation=recommendation,
                addendum=addendum, incidentalFindings=incidentalFindings,
                authors=authors, revision=revision
            )
        }
    }
    companion object {
        fun builder(): InterpretationPanelBuilder = InterpretationPanelBuilder()
        fun empty(): InterpretationPanel =  InterpretationPanel(
            null, null, null, null,
            null, null, null,
            null,
            null, null,
            emptyList(),
            null,
            null, null,
            emptyList(), false
        )
    }

    data class Variant(
        val snv: String?,
        val analysis: String?,
        val gene: String,
        val hgvsc: String?,
        val hgvsp: String?,
        @JsonProperty("origin_hgvsc")
        val originHgvsc: String?,
        @JsonProperty("origin_hgvsp")
        val originHgvsp: String?,
        val zygosity: String?,
        val disease: String?,
        val diseaseFullName: String?,
        val inheritance: String?,
        val interpretation: String?,
        val genPhenDb: String?,
        @JsonProperty("class")
        val clazz: Clazz
    )
    data class Author (
        val name: String,
        val comment: String?
    )
}
